src/helper/useBlockly.js

```diff
@@ -27,22 +30,74 @@ export function useBlockly() {
         return state.BallReducer;
     });
 
+    let lastBlockType;
+
+    const [workspaceCodeInterpreter, setWorkspaceCodeInterpreter] = useState();
+
     /**
      * Blockly-method to generate the code from the current workspace, print it to console (just for debugging)
      * and then executing the code.
      */
     const generateCode = () => {
+        BlocklyJS.STATEMENT_PREFIX = 'highlightBlock(%1);\n';
+        BlocklyJS.addReservedWords('highlightBlock');
+
         var code = BlocklyJS.workspaceToCode(
             simpleWorkspace.current.workspace
         );
-        //console.log(code);
-        try {
-            eval(code);
-        } catch (e) {
-            alert(e);
+
+        // remove the highlight from all the last execution
+        simpleWorkspace.current.workspace.highlightBlock(null);
+
+        const initApi = ((workspace) => {
+          return (interpreter, globalObject) => {
+            interpreter.setProperty(globalObject, 'highlightBlock', interpreter.createNativeFunction((id) => highlightBlock(workspace, id)));
+            interpreter.setProperty(globalObject, 'addRobot', interpreter.createNativeFunction(addRobot));
+            interpreter.setProperty(globalObject, 'moveRobot', interpreter.createNativeFunction(moveRobot));
+            interpreter.setProperty(globalObject, 'ballKick', interpreter.createNativeFunction(ballKick));
+            interpreter.setProperty(globalObject, 'turnRobot', interpreter.createNativeFunction(turnRobot));
+            interpreter.setProperty(globalObject, 'moveForward', interpreter.createNativeFunction(moveForward));
+          }
+        })(simpleWorkspace.current.workspace);
+
+        const Interpreter = window["Interpreter"];
+        console.log(code);
+        const myInterpreter = new Interpreter(code, initApi);
+        setWorkspaceCodeInterpreter(myInterpreter);
+
+    };
+
+    const interpret = () => {
+        //console.log("I'm thinking. Hard.");
+        if(workspaceCodeInterpreter) {
+            const nextStep = (robotList) => {
+                // TODO: hier können wir einfach den "type" vom *letzten* block uns
+                // angucken, und wenn der zB ein moveForward block war, dann müssen
+                // wir halt warten, bis der Roboter am Ziel angekommen ist. für die
+                // meisten anderen blöcke, sollte es ausreichen, wenn wir paar ms
+                // warten
+                //console.log("lastBlockType", lastBlockType);
+                const rob = robotList[0];
+                //console.log(rob.isActive);
+                if (rob.isActive) {
+                    return;
+                }
+                workspaceCodeInterpreter.step();
+                workspaceCodeInterpreter.step();
```

Comment:
I'm unsure whether there is *always* a highlight function after each function in the generated code (e.g. in the case of loops). My approach was to call `.step` only once and divide the interval by two. This way we are sure to never step over non-highlighting block.

Reply:
Wasn't actually necessary to do that, I took the second call out (was for testing and I forgot to remove it again)